packages/pds/src/migrate-script/index.ts

```diff
@@ -0,0 +1,417 @@
+import { createDeferrable } from '@atproto/common'
+import { envToCfg, envToSecrets, readEnv } from '../config'
+import AppContext from '../context'
+import { FailedTakedown, MigrateDb, Status, TransferPhase, getDb } from './db'
+
+type PdsInfo = {
+  id: number
+  did: string
+  url: string
+  agent: AtpAgent
+}
+
+export const runScript = async () => {
+  const db = await getDb('/data/migrate.db')
+  const env = readEnv()
+  const cfg = envToCfg(env)
+  const secrets = envToSecrets(env)
+  const ctx = await AppContext.fromConfig(cfg, secrets)
+  const adminToken = ui8.toString(
+    ui8.fromString(`admin:${secrets.adminPassword}`, 'utf8'),
+    'base64pad',
+  )
+  const pdsRes = await ctx.db.db.selectFrom('pds').selectAll().execute()
+  const pdsInfos = pdsRes.map((row) => ({
+    id: row.id,
+    did: row.did,
+    url: `https://${row.host}`,
+    agent: new AtpAgent({ service: `https://${row.host}` }),
+  }))
+  const todo = await db
+    .selectFrom('status')
+    .where('status.phase', '<', 7)
+    .where('failed', '!=', 1)
+    .selectAll()
+    .execute()
+  let pdsCounter = 0
+  for (const status of todo) {
+    let pdsId = status.pdsId
+    if (!pdsId) {
+      pdsId = pdsCounter % pdsInfos.length
+      pdsCounter++
+    }
+    const pdsInfo = pdsInfos[pdsId]
+    try {
+      await migrateRepo(ctx, db, pdsInfo, status, adminToken)
+      console.log(`completed migrating: ${status.did}`)
+    } catch (err) {
+      await db
+        .updateTable('status')
+        .set({ failed: 1 })
+        .where('did', '=', status.did)
+        .execute()
+      console.error(`failed to migrate: ${status.did}`, err)
+    }
+    pdsCounter++
+  }
+  console.log('DONE WITH ALL')
+}
+
+const migrateRepo = async (
+  ctx: AppContext,
+  db: MigrateDb,
+  pds: PdsInfo,
+  status: Status,
+  adminToken: string,
+) => {
+  console.log(`reserve (${status.did}, ${status.phase})`)
+  if (status.phase < TransferPhase.reservedKey) {
+    const signingKey = await reserveSigningKey(pds, status.did)
+    status.signingKey = signingKey
+    status.phase = TransferPhase.reservedKey
+    await updateStatus(db, status)
+  }
+
+  console.log(`import (${status.did}, ${status.phase})`)
+  if (status.phase < TransferPhase.initImport) {
+    const importedRev = await doImport(ctx, db, pds, status.did)
+    if (importedRev) {
+      status.importedRev = importedRev
+    }
+    status.phase = TransferPhase.initImport
+    await updateStatus(db, status)
+  }
+
+  console.log(`transfer (${status.did}, ${status.phase})`)
+  if (status.phase < TransferPhase.transferred) {
+    const importedRev = await lockAndTransfer(ctx, db, pds, status)
+    status.importedRev = importedRev
+    status.phase = TransferPhase.transferred
+    await updateStatus(db, status)
+  }
+
+  console.log(`prefs (${status.did}, ${status.phase})`)
+  if (status.phase < TransferPhase.preferences) {
+    try {
+      await transferPreferences(ctx, pds, status.did)
```

Comment:
May want to remove the entry from `failed_pref` if it succeeds here on the second go.

Reply:
Ah, actually I suppose we'll want a separate routine for handling the failures since `status.phase` still moves along during a failure 👍